orttraining/orttraining/core/graph/mixed_precision_transformer.cc

```diff
@@ -415,6 +415,28 @@ Status TransformGraphForMixedPrecision(Graph& graph,
     graph.AddInitializedTensor(weight_tensor_proto);
   }
 
+  // After FP16 initializers are created and added, update the graph's initializer list,
+  // replace the old fp32 initializer with the newly created one and keep it in-sync.
+  graph.SyncWithFP16Initializer(fp32_weight_name_to_fp16_node_arg_result);
+
+  // Handle pipeline case
+  for (auto& node : graph.Nodes()) {
+    // For send and recv node, if the tensor being sent or received is FP32, update its
```

Comment:
Not sure whether this always hold? (do we need to check producer/consumer of op??)
But I think it might be ok for test cases now.

Reply:
once the element type is updated in send/recv, the following consumer nodes will get populated later in graph.resolve(). The producer of recv is wait, which just pass through the data regardless its type.

---
In reply to: [434131796](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434131796) [](ancestors = 434131796)

Reply:
that is not my question. My question is that whether all send/recv can be always changed to lower-precision, fp16. Maybe in some case, our cut make a pipeline boundary of a tensor that requires high-precision data for convergence in mixed precision training. Then, this statement "For send and recv node, if the tensor being sent or received is FP32, update its attribute to FP16" won't hold, right? We might want to sure the same rule specifying FP16 or FP32. If it is OpType-based, since one of producer or consumer op is in another stage of graph, so we still need to check them, something like that. I think this can be future work.

Reply:
I see your point. I think the TransformStage2 will check those special ops and convert their input (adding cast) from fp32 back to fp16. 

---
In reply to: [434788808](https://github.com/microsoft/onnxruntime/pull/4113#discussion_r434788808) [](ancestors = 434788808)